group/assignment5/AlphabetFromDictionary.cpp

```diff
@@ -1,120 +1,175 @@
-    this->alphabetCharacters.clear();
-    vector<bool> isCharInDictionary(NUM_OF_NATIVE_CHARS, false);
-    for (string &word : this->dictionary) {
-        for (char c : word) {
-            if (!isCharInDictionary[c]) {
-                isCharInDictionary[c] = true;
-                this->alphabetCharacters.push_back(c);
-            }
-        }
-    }
+	this->alphabetCharacters.clear();
+	vector<bool> isCharInDictionary(NUM_OF_NATIVE_CHARS, false);
+	for (auto word : this->dictionary) {
+		for (char c : word) {
+			if (!isCharInDictionary[c]) {
+				isCharInDictionary[c] = true;
+				this->alphabetCharacters.push_back(c);
+			}
+		}
+	}
 }
 
-int AlphabetFromDictionary::findFirstMismatchIndex(const string &str1, const string &str2) {
-    int index = 0;
-    while (index < str1.size() && index < str2.size() && str1[index] == str2[index])
-        index++;
-    return (index < str1.size() && index < str2.size()) ? index : -1;
+int AlphabetFromDictionary::findFirstMismatchIndex(string &str1, string &str2) {
+	int index = 0;
+	while (index < str1.size() && index < str2.size() && str1[index] == str2[index])
+		index++;
+	return (index < str1.size() && index < str2.size()) ? index : -1;
 }
 
 void AlphabetFromDictionary::createCharPriorityGraph() {
-    int mismatchIndex;
-    string lastNonEmptyWord;
-
-    // find first non-empty word in dictionary
-    for (string &word : this->dictionary) {
-        if (!word.empty()) {
-            lastNonEmptyWord = word;
-            break;
-        }
-    }
-
-    for (string &word : this->dictionary) {
-        if (word.empty()) {
-            continue;
-        }
-        mismatchIndex = findFirstMismatchIndex(lastNonEmptyWord, word);
-        if (mismatchIndex != -1) {
-            char priorChar = lastNonEmptyWord[mismatchIndex], followingChar = word[mismatchIndex];
-            if (charPriorityGraph[followingChar][priorChar]) { // cycle in graph
-                this->isGraphCyclic = true;
-                this->orderedAlphabet = this->alphabetCharacters;
-                return;
-            }
-            this->charPriorityGraph[priorChar][followingChar] = true;
-        }
-        lastNonEmptyWord = word;
-    }
+	int mismatchIndex;
+	string lastNonEmptyWord;
+
+	// find first non-empty word in dictionary
+	for (string &word : this->dictionary) {
+		if (!word.empty()) {
+			lastNonEmptyWord = word;
+			break;
+		}
+	}
+
+	for (string &word : this->dictionary) {
+		if (word.empty()) {
+			continue;
+		}
+		mismatchIndex = findFirstMismatchIndex(lastNonEmptyWord, word);
+		if (mismatchIndex != -1) {
+			char priorChar = lastNonEmptyWord[mismatchIndex], followingChar = word[mismatchIndex];
+			if (charPriorityGraph[followingChar][priorChar]) { // cycle in graph
+				this->isGraphCyclic = true;
+				this->orderedAlphabet = this->alphabetCharacters;
+				return;
+			}
+			this->charPriorityGraph[priorChar][followingChar] = true;
+			this->inDegree[followingChar]++;

```

Comment:
There should be an if here:
_if (!this->charPriorityGraph[priorChar][followingChar]) {
    this->inDegree[followingChar]++;
}_
Otherwise we can count inDegree twice for the same priorChar. For example, in this dictionary: {"ART", "RAT", "CA", "CR"}, we will see that A comes before R twice, and inDegree of R will be 2 instead of 1